# Shervin's lab notebook (Ecological Genomics BIOL 381) 

## Author: Shervin Razavi
### Affiliation:  The University of Vermont
### E-mail contact: arazavi1@uvm.edu

### Start Date: 2020-01-13
### End Date: 2020-05-08
### Project Descriptions:   





# Table of Contents:   
* [Entry 1: 2020-01-13, Monday](#id-section1)
* [Entry 2: 2020-01-14, Tuesday](#id-section2)
* [Entry 3: 2020-01-15, Wednesday](#id-section3)
* [Entry 4: 2020-01-16, Thursday](#id-section4)
* [Entry 5: 2020-01-17, Friday](#id-section5)
* [Entry 6: 2020-01-20, Monday](#id-section6)
* [Entry 7: 2020-01-21, Tuesday](#id-section7)
* [Entry 8: 2020-01-22, Wednesday](#id-section8)
* [Entry 9: 2020-01-23, Thursday](#id-section9)
* [Entry 10: 2020-01-24, Friday](#id-section10)
* [Entry 11: 2020-01-27, Monday](#id-section11)
* [Entry 12: 2020-01-28, Tuesday](#id-section12)
* [Entry 13: 2020-01-29, Wednesday](#id-section13)
* [Entry 14: 2020-01-30, Thursday](#id-section14)
* [Entry 15: 2020-01-31, Friday](#id-section15)
* [Entry 16: 2020-02-03, Monday](#id-section16)
* [Entry 17: 2020-02-04, Tuesday](#id-section17)
* [Entry 18: 2020-02-05, Wednesday](#id-section18)
* [Entry 19: 2020-02-06, Thursday](#id-section19)
* [Entry 20: 2020-02-07, Friday](#id-section20)
* [Entry 21: 2020-02-10, Monday](#id-section21)
* [Entry 22: 2020-02-11, Tuesday](#id-section22)
* [Entry 23: 2020-02-12, Wednesday](#id-section23)
* [Entry 24: 2020-02-13, Thursday](#id-section24)
* [Entry 25: 2020-02-14, Friday](#id-section25)
* [Entry 26: 2020-02-17, Monday](#id-section26)
* [Entry 27: 2020-02-18, Tuesday](#id-section27)
* [Entry 28: 2020-02-19, Wednesday](#id-section28)
* [Entry 29: 2020-02-20, Thursday](#id-section29)
* [Entry 30: 2020-02-21, Friday](#id-section30)
* [Entry 31: 2020-02-24, Monday](#id-section31)
* [Entry 32: 2020-02-25, Tuesday](#id-section32)
* [Entry 33: 2020-02-26, Wednesday](#id-section33)
* [Entry 34: 2020-02-27, Thursday](#id-section34)
* [Entry 35: 2020-02-28, Friday](#id-section35)
* [Entry 36: 2020-03-02, Monday](#id-section36)
* [Entry 37: 2020-03-03, Tuesday](#id-section37)
* [Entry 38: 2020-03-04, Wednesday](#id-section38)
* [Entry 39: 2020-03-05, Thursday](#id-section39)
* [Entry 40: 2020-03-06, Friday](#id-section40)
* [Entry 41: 2020-03-09, Monday](#id-section41)
* [Entry 42: 2020-03-10, Tuesday](#id-section42)
* [Entry 43: 2020-03-11, Wednesday](#id-section43)
* [Entry 44: 2020-03-12, Thursday](#id-section44)
* [Entry 45: 2020-03-13, Friday](#id-section45)
* [Entry 46: 2020-03-16, Monday](#id-section46)
* [Entry 47: 2020-03-17, Tuesday](#id-section47)
* [Entry 48: 2020-03-18, Wednesday](#id-section48)
* [Entry 49: 2020-03-19, Thursday](#id-section49)
* [Entry 50: 2020-03-20, Friday](#id-section50)
* [Entry 51: 2020-03-23, Monday](#id-section51)
* [Entry 52: 2020-03-24, Tuesday](#id-section52)
* [Entry 53: 2020-03-25, Wednesday](#id-section53)
* [Entry 54: 2020-03-26, Thursday](#id-section54)
* [Entry 55: 2020-03-27, Friday](#id-section55)
* [Entry 56: 2020-03-30, Monday](#id-section56)
* [Entry 57: 2020-03-31, Tuesday](#id-section57)
* [Entry 58: 2020-04-01, Wednesday](#id-section58)
* [Entry 59: 2020-04-02, Thursday](#id-section59)
* [Entry 60: 2020-04-03, Friday](#id-section60)
* [Entry 61: 2020-04-06, Monday](#id-section61)
* [Entry 62: 2020-04-07, Tuesday](#id-section62)
* [Entry 63: 2020-04-08, Wednesday](#id-section63)
* [Entry 64: 2020-04-09, Thursday](#id-section64)
* [Entry 65: 2020-04-10, Friday](#id-section65)
* [Entry 66: 2020-04-13, Monday](#id-section66)
* [Entry 67: 2020-04-14, Tuesday](#id-section67)
* [Entry 68: 2020-04-15, Wednesday](#id-section68)
* [Entry 69: 2020-04-16, Thursday](#id-section69)
* [Entry 70: 2020-04-17, Friday](#id-section70)
* [Entry 71: 2020-04-20, Monday](#id-section71)
* [Entry 72: 2020-04-21, Tuesday](#id-section72)
* [Entry 73: 2020-04-22, Wednesday](#id-section73)
* [Entry 74: 2020-04-23, Thursday](#id-section74)
* [Entry 75: 2020-04-24, Friday](#id-section75)
* [Entry 76: 2020-04-27, Monday](#id-section76)
* [Entry 77: 2020-04-28, Tuesday](#id-section77)
* [Entry 78: 2020-04-29, Wednesday](#id-section78)
* [Entry 79: 2020-04-30, Thursday](#id-section79)
* [Entry 80: 2020-05-01, Friday](#id-section80)
* [Entry 81: 2020-05-04, Monday](#id-section81)
* [Entry 82: 2020-05-05, Tuesday](#id-section82)
* [Entry 83: 2020-05-06, Wednesday](#id-section83)
* [Entry 84: 2020-05-07, Thursday](#id-section84)
* [Entry 85: 2020-05-08, Friday](#id-section85)


------    
<div id='id-section1'/>   


### Entry 1: 2020-01-13, Monday.   



------    
<div id='id-section2'/>   


### Entry 2: 2020-01-14, Tuesday.   



------    
<div id='id-section3'/>   


### Entry 3: 2020-01-15, Wednesday.   



------    
<div id='id-section4'/>   


### Entry 4: 2020-01-16, Thursday.   



------    
<div id='id-section5'/>   


### Entry 5: 2020-01-17, Friday.   



------    
<div id='id-section6'/>   


### Entry 6: 2020-01-20, Monday.   



------    
<div id='id-section7'/>   


### Entry 7: 2020-01-21, Tuesday.   



------    
<div id='id-section8'/>   


### Entry 8: 2020-01-22, Wednesday.   

New learned codes:

'''

git status
git add --all 
git status
git commit -m
git push 
'''


------    
<div id='id-section9'/>   


### Entry 9: 2020-01-23, Thursday.   



------    
<div id='id-section10'/>   


### Entry 10: 2020-01-24, Friday.   



------    
<div id='id-section11'/>   


### Entry 11: 2020-01-27, Monday.   



------    
<div id='id-section12'/>   


### Entry 12: 2020-01-28, Tuesday.   



------    
<div id='id-section13'/>   


### Entry 13: 2020-01-29, Wednesday.   


TO incude code as a block, use:
'''


'''



To include code in a line use ' ' example:
I am doing this blah blah blah with the command ' grep /data/' command



Red spruce project:
1-Geographically isolated populations in the NE.
2-We have to see how we can maximize the efficiency of the restoration process.
3-80,000 120bp probes designed to find
4-Paired end reads are more powerful because the beginning and the end of a sequence have match the pairs

# How to read input/output files.
-----------) you can visualize it through FastQC
------------------------------------------) you can 'trim' it with Trimmomatic
----------)trimmed.fastq file 
------------------------------------------) through the bwa program you can map (pr align) to the reference genome
-----------------------------------------) post-processing through samtools ( .bam file) also removes duplicates.

# interpretation of data
The names of the trees is in the following format:
species name_Number of mother tree- Forward or backward read(R1 oor R2)


the @ on the first line is where the gx file starts ( first line contains the machine information + barcode0

second line: base pair read
Third line: Seperator between the second line and third line
fourth iine : a code that represents the quality of the reads (it contains the Phred score)


probability of failure = 10 ^ -(phrescore)/10

This does not take into account errors in prior steps ( e.g. PCR)




###My population is  XPK          REMEMBER THIS IT IS VERY IMPORTANT

Codding:

Make sure to alwayas pull and referesh your local repository before you commit and push.

the following code is used to do so:
'''
git pull
git commit --all
git push

'''

# Other useful git commands:
'''
git status

'''


------    
<div id='id-section14'/>   


### Entry 14: 2020-01-30, Thursday.   



------    
<div id='id-section15'/>   


### Entry 15: 2020-01-31, Friday.   



------    
<div id='id-section16'/>   


### Entry 16: 2020-02-03, Monday.   



------    
<div id='id-section17'/>   


### Entry 17: 2020-02-04, Tuesday.   



------    
<div id='id-section18'/>   


### Entry 18: 2020-02-05, Wednesday.   

Today we will be focusing on mapping to a subset of the reference genome.

Notes:

For the mapping scripts, we have to give both the forward and reverse reads of each individual and map it to the reference genome.  This will generate a sam file which can further on be converted to a bam file that is readable.


The Process_bam.sh script converts the sam files to bam files, deletes the PCR duplicates and  converts the final output to index files.
 
 
 we use seperate for loops for the processes used in this script.
 
 
 It is impotant to note that the PCR duplicates are removed because they lead to a lot of false positive calls.


Make sure to copy file locations accurately.


The mypipeline.sh script includes the two other scripts so that the "mother" defined variables (found in the mypipeline.sh script) do not have to be signified over and over again.

To test if it is working, make sure you have the main variables loaded on the system so you can run the other two scripts from outside of the mypipeline.sh script.

Pciture of the code:


![code]{





------    
<div id='id-section19'/>   


### Entry 19: 2020-02-06, Thursday.   



------    
<div id='id-section20'/>   


### Entry 20: 2020-02-07, Friday.   



------    
<div id='id-section21'/>   


### Entry 21: 2020-02-10, Monday.   



------    
<div id='id-section22'/>   


### Entry 22: 2020-02-11, Tuesday.   



------    
<div id='id-section23'/>   


### Entry 23: 2020-02-12, Wednesday.   

# Sam file stuff:


the second number is the flag number which gives you information about the read ( you can look up the meaning of the "flag" number  in the internet).
The third number is the place where the contig mapped to.
The forth number is the left position of the read
The fifth number is the quality number ( the higher the better)


samtools flagstat was preformed on the *.sam files.


Depth of coverage :How many reads stack up against any single position in the reference genome.


Samtools t-view ---) is a lightweight viewer


Genotype likelihood takes into account all of the possibilities that exist and does not simply go with the most probable situation.


ANGSD is a pretty cool platform you will use often!



# Workflow of ANGSD---) 
first site frequency spectrum (SNP frequency).
2- estimate nucleotide diversities
3-estiemate fst between all populations or pairwise between sets of populations.
perfom a genetic PCA based on estimatio of the genetic covariance matrix ( this 


The final code is very interesting and has a lot of relevant genomic information 

look up the meaning of the ANGSD command options.

# Major allele : most frequent allele.


# New commands:
use the "tail" command to get to the end of the file.




------    
<div id='id-section24'/>   


### Entry 24: 2020-02-13, Thursday.   



------    
<div id='id-section25'/>   


### Entry 25: 2020-02-14, Friday.   



------    
<div id='id-section26'/>   


### Entry 26: 2020-02-17, Monday.   



------    
<div id='id-section27'/>   


### Entry 27: 2020-02-18, Tuesday.   



------    
<div id='id-section28'/>   


### Entry 28: 2020-02-19, Wednesday.   

Folded SFS is used for cases that we do not know their clear ancestry.




the bottleneck is inherited from the ancestor.

Because Tajima's D is bigger than 1, the populations are porbably undergoing the same selective pressure.


Small values of theta and pi, hence not a lot going on.

The species not growing ( it is tanking because of the positive value of Tajima's D)

next step:


ANGSD_mypop.sh:


Make sure for the final analysis to not have the 


fold 1 is used to indicate the usage of the folded sfs.

coding note: make sure not to use "\" on the last option of the ANGSD command


   which particular SNPs are contributing to the local adaptations??------------------) GWAs







------    
<div id='id-section29'/>   


### Entry 29: 2020-02-20, Thursday.   



------    
<div id='id-section30'/>   


### Entry 30: 2020-02-21, Friday.   



------    
<div id='id-section31'/>   


### Entry 31: 2020-02-24, Monday.   


Transcriptomics : -Experimentally hold/control environmental conditions constant to reveal genetically controlled differences of phenotypes at the molecular level.

Phenotypes: Characteristics of an organism that can be observed. Disease resistance/susceptibility (cellular level) or color (morphological scale).

Central dogma of molecular bioloy: DNA ------(transcription-----) RNA -----(Translation)------) Protein -------) Phenotype


Reverse Ecology ----) Like GWAS



mutations ( at the DNA level):-----) -coding regions (Change of amino acids) 
                  -splice site variation
                  -promoter enhancer- expression level.
                           -expression level
                           -timing of expression 
                           -cell/tissue typpe of expressed gene 
                           -conditions under which a gene is expressed  (e.g. ogranisms that have diverged in their diversity of heat                                   stress
Epigenetic modificaitons

                           
Post translational modifications ( at the Protein level)




P = G + E + (G*E)
Experimentally induce yout phenotype of interest and control the environment --------------) molecular underpinnings of phenotypes.


Factors :
         - Controlling the envionmental condiiton (common garden)
         - Treatments/Conditions that you  are interested in
         - Populations of interest
         - Tissue that is being sampled
         - Life histoory change
         - Transgenerational enviornment 
         - Sexual dimorphism ( Sex and Reproductive stage )
         
         
HOW can we compare different species 

Work flow :
           1-Careful experimental design
               -Questions Hypotheses ( it could be broad [*omics] or a priori hypotheses]
           2- Experiment
                - sample tissues/individuals
                - sample RNA - labile           mRNA
                -Extract, prep, sequence        3' tag 
           3- Process + analyze data
               Quality check ----) Clean ----) qualiy check (.fastq)
               map to ref + Extract count data
               Normalization of data  -----------)) generation of data mattrix.
           4- integrate: functional enrichment aalyses, network analyses
                        and you can integrate with other data types such as SNP and TajD, pi, microbiome, epigenetic data, proteomics,                            environment.       
------    
<div id='id-section32'/>   


### Entry 32: 2020-02-25, Tuesday.   



------    
<div id='id-section33'/>   


### Entry 33: 2020-02-26, Wednesday.   




Toay we will work on transcriptomics:

#### Experimental Design:
Ten maternal families total; sample labels are “POP_FAM”
ASC_06, BRU_05, ESC_01, XBM_07, NOR_02, CAM_02, JAY_02, KAN_04, LOL_02, MMF_13
Two Source Climates (SourceClim:
HotDry (5 fams): ASC_06, BRU_05, ESC_01, XBM_07, NOR_02
CoolWet (5 fams): CAM_02, JAY_02, KAN_04, LOL_02, MMF_13
Experimental Treatments (Trt):
Control: watered every day, 16:8 L:D photoperiod at 23C:17C temps
Heat: 16:8 L:D photoperiod at 35C:26C temps (50% increase in day and night temps over controls)
Heat+Drought: Heat plus complete water witholding
Three time periods (Day):
Harvested tissues on Days 0, 5, and 10
Extracted RNA from whole seedlings (root, stem, needle tissue)
Aimed for 5 biological reps per Trt x SourceClim x Day combo, but day5 had few RNA extractions that worked




Questions that we could answer given the experimental desiggn ( two origin pops, 3 treatments, 3 experimental days).

1- Do individuals from different sources have different responses to different enviornments or different time points?
  you  can make regression models with interaction variables. Basically, you can make the interactions between the three factors.

2- Do families from different source climates differ in gene expression?


3- Is there a transcriptome wide response to heat stress? Does this change when the additional stress of drought is added


4- Which specific genes are involved in the above response?5-Do DE genes or pathways show evidence of positive selection?

6- Association mapping.



My populations are  #NOR H and XBM C



The headcrop comman was used in the trimmomatic program to delete the first 12 base pairs.

You can se the effect comparing the fastqc of the cleaned reads and the normals reads.


For the overrepresented sequences, the fastqc gives you data on the percentage the reads have been found and also the length of it.




------    
<div id='id-section34'/>   


### Entry 34: 2020-02-27, Thursday.   



------    
<div id='id-section35'/>   


### Entry 35: 2020-02-28, Friday.   



------    
<div id='id-section36'/>   


### Entry 36: 2020-03-02, Monday.   

# W hat is differential expression?
## Differences in transcript abundance b/w experimental groups --------) Genes that are changing

Sample prep ---------------) RNA extraction -----------------) Library prep --------------) SequNencing -----------)
Quality control (fastQC and trimming) -----------) mapping and counting (Salmon)

# Normalization: 
data matrix/gene count table
It is important to normalize fr library size.
Library composition
Question: 
How do you acccount for/analyze the  global differences betweeen samples

Normalization
Counts per million(CPM)) - Depth
Transcripts per Kilobase million-----------) Depth and gene lenth
Fragments per Kilobase million ------------) Depth and gene length
Edge R -----) Depth, Gene size and library composition 
DESeq2 -----) Depth and composition 

Differential Expression:

You have to take into account false positives and false negatives

Independent filtering :
        Limiting the number of tests based on a cut-off (basemean)
        Benamin, Hochberg(BH)
              Adjusted p-value
                    Control False Positives
                    FDR-False 
                    
Generalized Linear Model (GLM)
-------) Negative binomial distribution
Waod test -----) p-value



# Visualization:
   Sequencing summary statistics
   barplos of total read counts ( You can do # null genes)
   
   After normalization you can do a normalized count analysis.
   
   You can also do PCA
   
   Gene heat mapping (Uses hierarchical clustering to have the genes/ samples that are similar closer together).
   
   Ven diagrams: You can split up up-regulated genes, down-regulated genes, etc...
   
   MA plots ----) Log of FC [full change] over log of mean expression.
   
   
   Volcano plots : on one access you have log [full change] vs -log [p-value]
   
   log FC scatter plot: Log FC of condition 1 vs Log FC condition 2 
   
   individual gene treatment response curves :
   
   Examples: Relative Full Change vs treatment
   
## Question
You keep talking about ratio differences and their importance. How about the absolute transcriptomic differences. What if we have the same ratios, but 1  tenth of the transcription????





# Enrichment Analysis:
Looking at gene sets from functional databases ( Gene Ontology [GO])
      ----) Molecular function      -----) Example Adenylne cyclase activity
      ----) Biological process      -----) Puridimine biosynthesis
      ----) Cell Compartment        -----) Ribozome
      
      Another databse: KEGG pathway:
      Genetic Information Processing 
      ------) Transcription
      ------) Splicosome 
      
   
      Met
      hods: 
             Overrepresentation analysis 
             DEG ----) 2200 (ALL)        24(gene set)             18  (gene set 20
             total -----) 15,000           75                     120
             ------------) 14% of the genome is differentially expresse/d.                 32%         15%      
             
             
------    
<div id='id-section37'/>   


### Entry 37: 2020-03-03, Tuesday.   



------    
<div id='id-section38'/>   


### Entry 38: 2020-03-04, Wednesday.   


Today's code:


for file in /data/project_data/RS_RNASeq/fastq/cleanreads/NOR*C*R1.cl.fq

do 

salmon quant -i /data/project_data/RS_RNASeq/ReferenceTranscriptome/Pabies_H27_index -l A -r file --validateMappings -o /data/project_data/RS_RNASeq/salmon/cleanedreads

done 





for file in /data/project_data/RS_RNASeq/fastq/cleanreads/XBM_07_H*R1.cl.fq

do

salmon quant -i /data/project_data/RS_RNASeq/ReferenceTranscriptome/Pabies_H27_index -l A -r file --validateMappings -o /data/project_data/RS_RNASeq/salmon/cleanedreads

done




# important notes to consider for the future:

1-Before starting, make sure your destinations are accurate.


2- Check for typos from the very beginnning 



Mapping rate = Percent of the reads in each of the files that mapped to the genome ( not the other way around!)



High confidence----) 70 % or higher

Median 30-70 % percent

low = >0%





One possibility is that there is not a lot of conifer specific genes on genebank. ( maybe that is why the mapping rate is so low)


The reads had the highest mapping rate with the wholle index file (cds) rather than the HC and the HC27 index files.


The salmon index command works with the gz files ( you can change the kmer flag ( -k ), if you make it smaller it might make it better but it didn't make it better in our investigations. 


The mapping rates between the --seqbias and the normal ones were identical for the cleaned reads because they are clean!!!!!!


The quant.sf file is where the main info is.
![Example of quant.sf data ](https://github.com/arazavi78/Ecological-Genomics/blob/master/quant.sf.PNG)



R code explanation:


library(tximportData)
library(tximport)

#locate the directory containing the files. 
dir <- "/data/project_data/RS_RNASeq/salmon/"
list.files(dir)

# read in table with sample ids
samples <- read.table("/data/project_data/RS_RNASeq/salmon/RS_samples.txt", header=TRUE)

# now point to quant files
all_files <- file.path(dir, samples$sample, "quant.sf")

#Here we are giving names to the samples
names(all_files) <- samples$sample

# what would be used if linked transcripts to genes
#here the tximport  
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)

# to be able to run without tx2gene
txi <- tximport(all_files, type = "salmon", txOut=TRUE)  
names(txi)

head(txi$counts)

countsMatrix <- txi$counts
dim(countsMatrix)
#[1] 66069    76

# To write out
write.table(countsMatrix, file = "RS_countsMatrix.txt", col.names = T, row.names = T, quote = F) 






 # Coding
------    
<div id='id-section39'/>   


### Entry 39: 2020-03-05, Thursday.   



------    
<div id='id-section40'/>   


### Entry 40: 2020-03-06, Friday.   



------    
<div id='id-section41'/>   


### Entry 41: 2020-03-09, Monday.   



------    
<div id='id-section42'/>   


### Entry 42: 2020-03-10, Tuesday.   



------    
<div id='id-section43'/>   


### Entry 43: 2020-03-11, Wednesday.   



------    
<div id='id-section44'/>   


### Entry 44: 2020-03-12, Thursday.   



------    
<div id='id-section45'/>   


### Entry 45: 2020-03-13, Friday.   



------    
<div id='id-section46'/>   


### Entry 46: 2020-03-16, Monday.   



------    
<div id='id-section47'/>   


### Entry 47: 2020-03-17, Tuesday.   



------    
<div id='id-section48'/>   


### Entry 48: 2020-03-18, Wednesday.   



------    
<div id='id-section49'/>   


### Entry 49: 2020-03-19, Thursday.   



------    
<div id='id-section50'/>   


### Entry 50: 2020-03-20, Friday.   



------    
<div id='id-section51'/>   


### Entry 51: 2020-03-23, Monday.   



------    
<div id='id-section52'/>   


### Entry 52: 2020-03-24, Tuesday.   



------    
<div id='id-section53'/>   


### Entry 53: 2020-03-25, Wednesday.   



------    
<div id='id-section54'/>   


### Entry 54: 2020-03-26, Thursday.   



------    
<div id='id-section55'/>   


### Entry 55: 2020-03-27, Friday.   



------    
<div id='id-section56'/>   


### Entry 56: 2020-03-30, Monday.   



------    
<div id='id-section57'/>   


### Entry 57: 2020-03-31, Tuesday.   



------    
<div id='id-section58'/>   


### Entry 58: 2020-04-01, Wednesday.   



------    
<div id='id-section59'/>   


### Entry 59: 2020-04-02, Thursday.   



------    
<div id='id-section60'/>   


### Entry 60: 2020-04-03, Friday.   



------    
<div id='id-section61'/>   


### Entry 61: 2020-04-06, Monday.   



------    
<div id='id-section62'/>   


### Entry 62: 2020-04-07, Tuesday.   



------    
<div id='id-section63'/>   


### Entry 63: 2020-04-08, Wednesday.   



------    
<div id='id-section64'/>   


### Entry 64: 2020-04-09, Thursday.   



------    
<div id='id-section65'/>   


### Entry 65: 2020-04-10, Friday.   



------    
<div id='id-section66'/>   


### Entry 66: 2020-04-13, Monday.   



------    
<div id='id-section67'/>   


### Entry 67: 2020-04-14, Tuesday.   



------    
<div id='id-section68'/>   


### Entry 68: 2020-04-15, Wednesday.   



------    
<div id='id-section69'/>   


### Entry 69: 2020-04-16, Thursday.   



------    
<div id='id-section70'/>   


### Entry 70: 2020-04-17, Friday.   



------    
<div id='id-section71'/>   


### Entry 71: 2020-04-20, Monday.   



------    
<div id='id-section72'/>   


### Entry 72: 2020-04-21, Tuesday.   



------    
<div id='id-section73'/>   


### Entry 73: 2020-04-22, Wednesday.   



------    
<div id='id-section74'/>   


### Entry 74: 2020-04-23, Thursday.   



------    
<div id='id-section75'/>   


### Entry 75: 2020-04-24, Friday.   



------    
<div id='id-section76'/>   


### Entry 76: 2020-04-27, Monday.   



------    
<div id='id-section77'/>   


### Entry 77: 2020-04-28, Tuesday.   



------    
<div id='id-section78'/>   


### Entry 78: 2020-04-29, Wednesday.   



------    
<div id='id-section79'/>   


### Entry 79: 2020-04-30, Thursday.   



------    
<div id='id-section80'/>   


### Entry 80: 2020-05-01, Friday.   



------    
<div id='id-section81'/>   


### Entry 81: 2020-05-04, Monday.   



------    
<div id='id-section82'/>   


### Entry 82: 2020-05-05, Tuesday.   



------    
<div id='id-section83'/>   


### Entry 83: 2020-05-06, Wednesday.   



------    
<div id='id-section84'/>   


### Entry 84: 2020-05-07, Thursday.   



------    
<div id='id-section85'/>   


### Entry 85: 2020-05-08, Friday.   



